### ISSUE TITLE
Reciprocal and related functions have poor precision.

### ISSUE THREAD

**Issue by jasondavies** [tenstorrent/tt-metal#21622]
### Describe the bug

The `_calculate_reciprocal_` LLK function has poor precision due to a low quality initial estimate, which affects various other functions that rely on it.

See tenstorrent/tt-llk#154 for report and tenstorrent/tt-llk#155 for proposed fix.

As part of fixing this issue, I noticed that several places in `tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/` use hardcoded constants instead of calling the LLK initialisation function to set these constants.

In order to use updated LLK with new constants, these hardcoded constants need to be removed and ideally replaced by calls to the appropriate LLK init functions.

**Comment by adifrancescoTT** [comment]
Thanks @jasondavies :) Hey @cmaryanTT , anyone in your team can look into this for estimate/timing? Thanks!

**Comment by cmaryanTT** [comment]
@KalaivaniMCW and team will look at this with high priority.

**Comment by cmaryanTT** [comment]
@jasondavies - I thought I saw something touching this in one of your PRs, though I can't find it now. Can you please confirm if you intend to make these changes or if you'd like us to make these changes.

**Comment by jasondavies** [comment]
@cmaryanTT I was hoping @Shubhamsaboo might promote this to a bounty.  I still intend to push through these changes, which include:

- #21623 - this is essentially preparing the way for the recip improvements I made in tt-llk.  (This actually got merged and then had to be reverted because of broken tests in tt-metal, now fixed.)
- #21892 - probably best if I implement my planned improvements to tt-llk first and then report my precision/performance improvements to this PR

My tt-llk PR unfortunately got reverted because of the broken tests in tt-metal (now fixed):

- tenstorrent/tt-llk#155

For ease of review, my plan is:

1. Open a new tt-llk PR with all of the recip/related function improvements.
2. Open or update the existing tt-metal PR reporting all of the improvements and this can be merged with the next tt-llk uplift.

**Comment by zzigler-tt** [comment]
@cmaryanTT can we remove you from being assigned to this bug if it was in-fact made into a bounty.

**Comment by cmaryanTT** [comment]
Done

**Comment by cmaryanTT** [comment]
Hi @jasondavies - can you please provide an update on this issue. From a quick glance, it looks like these PRs are nearly ready
https://github.com/tenstorrent/tt-metal/pull/21623
https://github.com/tenstorrent/tt-llk/pull/230

Do you need any assistance from us?

**Comment by jasondavies** [comment]
@cmaryanTT Thanks for following up.  These PRs are blocked at the moment due to unexplained PCC drops in SD and BERT tests.  In both cases it seems that the issue is due to the improved precision of rsqrt (note: I've replaced sqrt+recip with rsqrt where possible) for layernorm.

- BERT: https://github.com/tenstorrent/tt-metal/pull/21623#issuecomment-3084218426
- SD: https://github.com/tenstorrent/tt-metal/pull/21623#issuecomment-3057915083

In both cases the PCC drop is quite small, but it does seem noticeable in the SD image test.

For the BERT test, there is a PCC drop even on main if `ttnn.layer_norm` is patched to use PyTorch, which would indicate the real issue is elsewhere, and perhaps it was relying on some truncation/rounding behaviour in the original layernorm.

For the SD test, I haven't managed to patch `ttnn.layer_norm` in the same way to reproduce.  One way someone could help is to attempt to patch `ttnn.layer_norm` to use PyTorch and see if the same thing occurs.

Another idea is we could get these PRs merged by adding the old sqrt+recip for layernorm under a temporary API while we figure out what's wrong.

**Comment by cmaryanTT** [comment]
@jasondavies - thanks for the feedback, I'll speak with the layer_norm owner to see how we can proceed. There's an overhaul of layer_norm happening now, which may help.

**Comment by cmaryanTT** [comment]
@jasondavies - we reviewed this internally, and we think it's best if you commit these under a different name, and continue using the old version wherever it's problematic; and use the new version wherever it's fine. But this is win-win, you can close this issue and claim the bounty, we get to unblock a bunch of work that could benefit from the improved accuracy, and we'll take on the work of figuring out the problematic cases as a separate task. It's temporarily hacky, but it'll get cleaned up as a high priority item.

**Comment by jasondavies** [comment]
@cmaryanTT Great!  My suggestion would be to try and combine the old sqrt and recip into a single ln_rsqrt for use with the current layernorm implementation.  I _think_ that's the only thing that ran into issues.

**Comment by jasondavies** [comment]
It would be less work to introduce a template parameter for rsqrt_tile and rsqrt_tile_init, `bool layernorm_compat`, which can be removed later once the issue is resolved.

**Comment by cmaryanTT** [comment]
@KalaivaniMCW 's team is going to make the layernorm changes. @nathan-TT is going to do a deep dive into the layernorm problem, for which I'll make a sub-issue.

**Comment by KalaivaniMCW** [comment]
Observing a failure with one of the VIT tests when using fused sqrt+recip for Layernorm. 
we're debugging

<img width="1077" height="173" alt="Image" src="https://github.com/user-attachments/assets/22de2b74-85aa-4ddb-99aa-4494f2eb4845" />

**Comment by KalaivaniMCW** [comment]
@jasondavies  Do you have the link of the failed CI runs from before ?

**Comment by KalaivaniMCW** [comment]
Although `tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py` is passing with the fused op, not sure why that one test in VIT is showing slight drop( 0.876 < 0.88) . So far made changes in layernorm and groupnorm kernels

**Comment by jasondavies** [comment]
@KalaivaniMCW There's a link to the most recent CI runs in https://github.com/tenstorrent/tt-metal/pull/21623 but I think that only shows the first test failure listed below:

- `pytest 'models/demos/blackhole/stable_diffusion/tests/test_unet_2d_condition_model.py::test_unet_2d_condition_model_512x512[2-4-64-64-device_params=l1_small_size_24576]'`
- `pytest 'tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py::test_bert_for_question_answering[num_hidden_layers=1-sequence_size=128-batch_size=8-model_name=mrm8488/bert-tiny-finetuned-squadv2]'`
- `pytest models/demos/wormhole/stable_diffusion/tests/test_demo.py::test_demo_sd`

Note: some tests may need `WH_ARCH_YAML="wormhole_b0_80_arch_eth_dispatch.yaml"` on WH.

**Comment by jasondavies** [comment]
Just pushed my changes to both PRs with the addition of a new `bool layernorm_compat` template parameter for rsqrt_tile/rsqrt_tile_init.

The following tests now pass on WH (be1a526dfe0d976416dbfac8a67254b9719bd34b):

> * `pytest 'models/demos/blackhole/stable_diffusion/tests/test_unet_2d_condition_model.py::test_unet_2d_condition_model_512x512[2-4-64-64-device_params=l1_small_size_24576]'`
> * `pytest 'tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py::test_bert_for_question_answering[num_hidden_layers=1-sequence_size=128-batch_size=8-model_name=mrm8488/bert-tiny-finetuned-squadv2]'`
> * `pytest models/demos/wormhole/stable_diffusion/tests/test_demo.py::test_demo_sd`

**Comment by KalaivaniMCW** [comment]
The VIT failure is now resolved. We were able to get a clean APC. 
Testing model pipelines now to see if there are other failures. 

@jasondavies 
Do you still need this fused sqrt+recip merged ? Or do you have a workaround ?

**Comment by jasondavies** [comment]
@KalaivaniMCW I've actually already implemented the "old" sqrt+recip under a special template param, so I don't need anything merged, sorry I think we got our wires crossed there.

**Comment by jasondavies** [comment]
> The VIT failure is now resolved. We were able to get a clean APC.

Did you do anything specific to resolve this one?  I got the impression it was due to my higher precision recip implementation, so in the end I opted to reduce the PCC from 0.88 to 0.87.  But if there's a better resolution it would be great if you could share it.

**Comment by KalaivaniMCW** [comment]
@jasondavies 
I tried your changes, like moving the output of the sqrt operation to sfpi::dst_reg[0] instead of assigning it to a vFloat variable and then called the recip fn  [kalaivani/ln_rsqrt
](https://github.com/tenstorrent/tt-metal/compare/main...kalaivani/ln_rsqrt)

With this change the VIT test and TG test failure on APC resolved. 
The model CI is down - waiting for it to be resolved.

**Comment by jasondavies** [comment]
@KalaivaniMCW Makes sense - thanks.  Yes, I had to do that to emulate the exact behaviour of the old version, i.e. when dst is in 16-bit mode, assigning to sfpi::dst_reg[0] truncates the fp32 value before calculating the reciprocal.

The VIT failure I'm seeing is due to using the new recip in softmax, so I guess it's just a separate thing.  When I use approx_mode=False, I get 0.8798630377940498.

**Comment by KalaivaniMCW** [comment]
APC: https://github.com/tenstorrent/tt-metal/actions/runs/16668148465
SC: 
https://github.com/tenstorrent/tt-metal/actions/runs/16673236362
https://github.com/tenstorrent/tt-metal/actions/runs/16673760246
https://github.com/tenstorrent/tt-metal/actions/runs/16673768397

**Comment by jasondavies** [comment]
FYI @marty1885 @minaliuTT this was merged last week.
